### PR TITLE
FAISS HNSW range search

### PR DESCRIPTION
**Pull request by alexanderguzhva**
/kind enhancement
issue: #869

### CODE REVIEW

**Comment by alexanderguzhva** [comment]
rebase to master and add bitset-related unit tests

**Comment by alexanderguzhva** [comment]
/hold
problem with brute force for range search

**Comment by alexanderguzhva** [comment]
/unhold
A possible problem with both hnswlib & faiss: low recall rate with 0.9x filtering when brute-force did not get triggered

**Review by liliu-z — COMMENTED** [review]
We are planning to use iterator to implement RangeSearch to reduce the maintain overhead. And this is what we are doing in Cardinal.
Offline talked with @alwayslove2013, who works on that. The reason why Knowhere hasn't adapted this method for now is that the iterator implementation is not fully optimized yet. 

We can check this in for supporting Faiss-based HNSW for now. And consider the iterator-based RangeSearch in the future.

File: src/index/hnsw/impl/IndexBruteForceWrapper.cc
Comment (alexanderguzhva) [reply src/index/hnsw/impl/IndexBruteForceWrapper.cc:165]: I'd like to keep it as is for now, bcz otherwise I'll need to use C++-20 templates lambdas features. Besides, this code is simpler to debug.
Further refactoring will follow, I just need to get a functional version first.

File: thirdparty/faiss/faiss/cppcontrib/knowhere/IndexBruteForceWrapper.h
Comment (alexanderguzhva) [reply thirdparty/faiss/faiss/cppcontrib/knowhere/IndexBruteForceWrapper.h:37]: 3. oops :) I will correct this

**Review by liliu-z — APPROVED** [review]
/lgtm

**Comment by liliu-z** [comment]
@alwayslove2013 Plz help take a check about the Range Search logic

**Comment by liliu-z** [comment]
@alexanderguzhva Plz create a simple issue for this PR to unblock it

**Pull request merged**
